README.md
# ngx-carousel

Lightweight and simple carousel for angular.

`Note: This carousel doesn't include any css. go and customize css for buttons, items except ngxcarousel and ngxcarousel-inner`

## changelog [![NPM version](https://badge.fury.io/js/ngx-carousel.png)](http://badge.fury.io/js/ngx-carousel)

for ChangeLog go to [CHANGELOG.md](https://github.com/sheikalthaf/ngx-carousel/blob/master/CHANGELOG.md)

# Universal Version

**Angular Universal** Version available on [ngu-carousel](https://github.com/sheikalthaf/ngu-carousel)

## Demo

Demo available [Here](https://sheikalthaf.github.io/ngx-carousel)

## Installation

`npm install ngx-carousel --save`

Now ngx-carousel supports touch with the help of hammerjs

`npm install hammerjs --save`

## Sample

Include CarouselModule in your app module:

```javascript
import { NgxCarouselModule } from 'ngx-carousel';
import 'hammerjs';

@NgModule({
  imports: [
    NgxCarouselModule
  ],
})
export class AppModule { }
```

Then use in your component:

```javascript
import { Component } from '@angular/core';
import { NgxCarousel } from 'ngx-carousel';

@Component({
  selector: 'sample',
  template: `
    <ngx-carousel
        [inputs]="carouselOne"
        (carouselLoad)="myfunc($event)">
          <ngx-item NgxCarouselItem>
            ....
          </ngx-item>
          <ngx-item NgxCarouselItem>
            ....
          </ngx-item>
          <ngx-item NgxCarouselItem>
            ....
          </ngx-item>
          <button NgxCarouselPrev class='leftRs'>&lt;</button>
          <button NgxCarouselNext class='rightRs'>&gt;</button>
    </ngx-carousel>
  `,
})
export class SampleComponent implements OnInit {

  public carouselOne: NgxCarousel;

  ngOnInit() {
    this.carouselOne = {
      grid: {xs: 1, sm: 1, md: 1, lg: 1, all: 0},
      slide: 1,
      speed: 400,
      interval: 4000,
      point: {
        visible: true
      },
      load: 2,
      touch: true,
      loop: true,
      custom: 'banner'
    }
  }

  public myfunc(event: Event) {
     // carouselLoad will trigger this funnction when your load value reaches
     // it is helps to load the data by parts to increase the performance of the app
     // must use feature to all carousel
  }


}
```

## Input Interface

```javascript

export class NgxCarouselStore {
  type: string;
  deviceType: DeviceType;
  classText: string;
  items: number;
  load: number;
  deviceWidth: number;
  carouselWidth: number;
  itemWidth: number;
  visibleItems: ItemsControl;
  slideItems: number;
  itemWidthPer: number;
  itemLength: number;
  currentSlide: number;
  easing: string;
  speed: number;
  transform: Transfrom;
  loop: boolean;
  dexVal: number;
  touchTransform: number;
  touch: Touch;
  isEnd: boolean;
  isFirst: boolean;
  isLast: boolean;
}
export type DeviceType = 'xs' | 'sm' | 'md' | 'lg' | 'all';

export class ItemsControl {
  start: number;
  end: number;
}

export class Touch {
  active?: boolean;
  swipe: string;
  velocity: number;
}

export class NgxCarousel {
  grid: Transfrom;
  slide?: number;
  speed?: number;
  interval?: number;
  animation?: Animate;
  point?: Point;
  type?: string;
  load?: number;
  custom?: Custom;
  loop?: boolean;
  touch?: boolean;
  easing?: string;
}

export class Grid {
  xs: number;
  sm: number;
  md: number;
  lg: number;
  all: number;
}

export interface Point {
  visible: boolean;
  pointStyles?: string;
}

export type Custom = 'banner';
export type Animate = 'lazy';


```

| Command | Type | Required | Description |
| --- | --- | --- | --- |
| `grid` | Object | Yes | **xs** - mobile, **sm** - tablet, **md** - desktop, **lg** - large desktops, **all** - fixed width (When you use **all** make others 0 and vise versa) |
| `slide` | number | optional | It is used to slide the number items on click |
| `speed` | milli seconds | optional | It is used for time taken to slide the number items |
| `interval` | milli seconds | optional | It is used to make carousel auto slide with given value. interval defines the interval between slides |
| `load` | number | optional | is used to load the items similar to pagination. the carousel will tigger the carouslLoad function to load another set of items. it will help you to improve the performance of the app.**`(carouselLoad)="myfunc($event)"`** |
| `point.visible` | boolean | optional | It is used to indicate no. of slides and also shows the current active slide. |
| `point.pointStyle` | string | optional | It is used to customize the point indicator. make sure your check the guide. |
| `touch` | boolean | optional | It is used to active touch support to the carousel. |
| `easing` | string | optional | It is used to define the easing style of the carousel. Only define the ease name without any timing like `ease`,`ease-in` |
| `loop` | boolean | optional | It is used to loop the `ngx-item | ngx-tile`. It must be true for `interval` |
| `animation` | string | optional | It is used to animate the sliding items. currently it only supports `lazy`. more coming soon and also with custom css animation option |
| `custom` | string | optional | It is you to define the purpose of the carousel. currently it only supports `banner`. more coming soon and also with custom css animation option |

### Custom css for Point

```html
<ul class="ngxcarouselPoint">
    <li *ngFor="let i of pointNumbers; let i = index" [class.active]="i==pointers"></li>
</ul>
```

This is HTML I'm using in the carousel. Add your own css according to this elements in `pointStyles`. check below guide for more details.

```html

<ngx-carousel
      [inputs]="carouselBanner"
      [moveToSlide]="2"
      (onMove)="onmoveFn($event)"
      (carouselLoad)="loadItemsFn()"
      (afterCarouselViewed)="afterCarouselViewedFn($event)">
</ngx-carousel>

```

* `inputs` is an `Input` and It accepts `NgxCarousel`.
* `moveToSlide` is an `Input` which accepts point numbers. Numbers represents no of slide to be done.
* `onMove` is an `Output` which triggered on every slide before start and it will emit `$event` as `NgxCarouselStore` object.
* `carouselLoad` is an `Output` which triggered when slide reaches the end on items based on inputs `load`.
* `afterCarouselViewed` is an `Output` which triggered after carousel viewed and it will emit `$event` as `NgxCarouselStore` object.

# Getstarted guide

## Banner with Custom point style

```javascript
import { Component } from '@angular/core';
import { NgxCarousel, NgxCarouselStore } from 'ngx-carousel';

@Component({
  selector: 'app-carousel',
  template: `
    <ngx-carousel
      [inputs]="carouselBanner"
      [moveToSlide]="1"
      (onMove)="onmoveFn($event)"
      (afterCarouselViewed)="afterCarouselViewedFn($event)">

          <ngx-item NgxCarouselItem class="bannerStyle">
              <h1>1</h1>
          </ngx-item>

          <ngx-item NgxCarouselItem class="bannerStyle">
              <h1>2</h1>
          </ngx-item>

          <ngx-item NgxCarouselItem class="bannerStyle">
              <h1>3</h1>
          </ngx-item>

          <button NgxCarouselPrev class='leftRs'>&lt;</button>
          <button NgxCarouselNext class='rightRs'>&gt;</button>
    </ngx-carousel>
  `,
  styles: [`
    .bannerStyle h1 {
        background-color: #ccc;
        min-height: 300px;
        text-align: center;
        line-height: 300px;
    }
    .leftRs {
        position: absolute;
        margin: auto;
        top: 0;
        bottom: 0;
        width: 50px;
        height: 50px;
        box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);
        border-radius: 999px;
        left: 0;
    }

    .rightRs {
        position: absolute;
        margin: auto;
        top: 0;
        bottom: 0;
        width: 50px;
        height: 50px;
        box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);
        border-radius: 999px;
        right: 0;
    }
  `]
})
export class Sample implements OnInit {


  ngOnInit(){

    this.carouselBanner = {
      grid: {xs: 1, sm: 1, md: 1, lg: 1, all: 0},
      slide: 1,
      speed: 400,
      interval: 4000,
      point: {
        visible: true,
        pointStyles: `
          .ngxcarouselPoint {
            list-style-type: none;
            text-align: center;
            padding: 12px;
            margin: 0;
            white-space: nowrap;
            overflow: auto;
            position: absolute;
            width: 100%;
            bottom: 20px;
            left: 0;
            box-sizing: border-box;
          }
          .ngxcarouselPoint li {
            display: inline-block;
            border-radius: 999px;
            background: rgba(255, 255, 255, 0.55);
            padding: 5px;
            margin: 0 3px;
            transition: .4s ease all;
          }
          .ngxcarouselPoint li.active {
              background: white;
              width: 10px;
          }
        `
      },
      load: 2,
      loop: true,
      touch: true
    }
  }

  /* This will be triggered after carousel viewed */
  afterCarouselViewedFn(data) {
    console.log(data);
  }

  /* It will be triggered on every slide*/
  onmoveFn(data: NgxCarouselStore) {
    console.log(data);
  }

}

```

## Tile

```javascript
import { Component } from '@angular/core';
import { NgxCarousel } from 'ngx-carousel';

@Component({
  selector: 'app-carousel',
  template: `
    <ngx-carousel
      [inputs]="carouselTile"
      (carouselLoad)="carouselTileLoad($event)">

            <ngx-tile NgxCarouselItem *ngFor="let Tile of carouselTileItems">
                <h1>{{Tile + 1}}</h1>
            </ngx-tile>

          <button NgxCarouselPrev class='leftRs'>&lt;</button>
          <button NgxCarouselNext class='rightRs'>&gt;</button>
    </ngx-carousel>
  `,
  styles: [`

    h1{
      min-height: 200px;
      background-color: #ccc;
      text-align: center;
      line-height: 200px;
    }
    .leftRs {
        position: absolute;
        margin: auto;
        top: 0;
        bottom: 0;
        width: 50px;
        height: 50px;
        box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);
        border-radius: 999px;
        left: 0;
    }

    .rightRs {
        position: absolute;
        margin: auto;
        top: 0;
        bottom: 0;
        width: 50px;
        height: 50px;
        box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);
        border-radius: 999px;
        right: 0;
    }
  `]
})
export class Sample implements OnInit {

  public carouselTileItems: Array<any>;
  public carouselTile: NgxCarousel;

  ngOnInit(){
    this.carouselTileItems = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];

    this.carouselTile = {
      grid: {xs: 2, sm: 3, md: 3, lg: 5, all: 0},
      slide: 2,
      speed: 400,
      animation: 'lazy',
      point: {
        visible: true
      },
      load: 2,
      touch: true,
      easing: 'ease'
    }
  }

  public carouselTileLoad(evt: any) {

    const len = this.carouselTileItems.length
    if (len <= 30) {
      for (let i = len; i < len + 10; i++) {
        this.carouselTileItems.push(i);
      }
    }

  }

     // carouselLoad will trigger this funnction when your load value reaches
     // it is helps to load the data by parts to increase the performance of the app
     // must use feature to all carousel

}
```

## Tile with custom point style

```javascript
import { Component } from '@angular/core';
import { NgxCarousel } from 'ngx-carousel';

@Component({
  selector: 'app-carousel',
  template: `
    <ngx-carousel
      [inputs]="carouselTile"
      (carouselLoad)="carouselTileLoad($event)">

            <ngx-tile NgxCarouselItem *ngFor="let Tile of carouselTileItems">
                <h1>{{Tile + 1}}</h1>
            </ngx-tile>

          <button NgxCarouselPrev class='leftRs'>&lt;</button>
          <button NgxCarouselNext class='rightRs'>&gt;</button>
    </ngx-carousel>
  `,
  styles: [`

    h1{
      min-height: 200px;
      background-color: #ccc;
      text-align: center;
      line-height: 200px;
    }
    .leftRs {
        position: absolute;
        margin: auto;
        top: 0;
        bottom: 0;
        width: 50px;
        height: 50px;
        box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);
        border-radius: 999px;
        left: 0;
    }

    .rightRs {
        position: absolute;
        margin: auto;
        top: 0;
        bottom: 0;
        width: 50px;
        height: 50px;
        box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);
        border-radius: 999px;
        right: 0;
    }
  `]
})
export class Sample implements OnInit {

  public carouselTileItems: Array<any>;
  public carouselTile: NgxCarousel;

  ngOnInit(){
    this.carouselTileItems = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];

    this.carouselTile = {
      grid: {xs: 2, sm: 3, md: 3, lg: 5, all: 0},
      slide: 2,
      speed: 400,
      animation: 'lazy',
      point: {
        visible: true,
        pointStyles: `
          .ngxcarouselPoint {
            list-style-type: none;
            text-align: center;
            padding: 12px;
            margin: 0;
            white-space: nowrap;
            overflow: auto;
            box-sizing: border-box;
          }
          .ngxcarouselPoint li {
            display: inline-block;
            border-radius: 50%;
            border: 2px solid rgba(0, 0, 0, 0.55);
            padding: 4px;
            margin: 0 3px;
            transition-timing-function: cubic-bezier(.17, .67, .83, .67);
            transition: .4s;
          }
          .ngxcarouselPoint li.active {
              background: #6b6b6b;
              transform: scale(1.2);
          }
        `
      },
      load: 2,
      touch: true,
      easing: 'ease'
    }
  }

  public carouselTileLoad(evt: any) {

    const len = this.carouselTileItems.length
    if (len <= 30) {
      for (let i = len; i < len + 10; i++) {
        this.carouselTileItems.push(i);
      }
    }

  }

     // carouselLoad will trigger this funnction when your load value reaches
     // it is helps to load the data by parts to increase the performance of the app
     // must use feature to all carousel

}
```

## License

[MIT](LICENSE) license.

CHANGELOG.md
# changelog [![NPM version](https://badge.fury.io/js/ngx-carousel.png)](http://badge.fury.io/js/ngx-carousel) 

### Before Upgrading once read the [Readme.md](https://github.com/sheikalthaf/ngx-carousel/blob/master/README.md) file

## 1.3.2

### Features

* `afterCarouselViewed` is an `Output` which triggered after carousel viewed and it will emit `$event` as `NgxCarouselStore` object [#47](https://github.com/sheikalthaf/ngx-carousel/issues/47)

### Issues fixed

* fixed touch when carousel scrolled vertically by [matt](https://github.com/iamlothian) [#20](https://github.com/sheikalthaf/ngx-carousel/issues/20)

* Fixed the nested carousel issue by [matt](https://github.com/iamlothian) [#42](https://github.com/sheikalthaf/ngx-carousel/issues/42)

* Fixed the source fails issue by [matt](https://github.com/iamlothian) [matt](https://github.com/iamlothian) [#24](https://github.com/sheikalthaf/ngx-carousel/issues/24)

## 1.3.0

### Features

* `moveToSlide` is an `Input` which accepts point numbers. Numbers represents no of slide to be done. You can change it dynamically and slider will based on the number.

* `onMove` is an `Output` which triggered on every slide before start and it will emit `$event` as `NgxCarouselStore` object.

### Issues fixed

* Added feature to move slider to specific point number[#21](https://github.com/sheikalthaf/ngx-carousel/issues/21)

* Fixed the swipe issue on iphone 6[#22](https://github.com/sheikalthaf/ngx-carousel/issues/22)

* Fixed the point issue [#24](https://github.com/sheikalthaf/ngx-carousel/issues/24)

* Added to `onMove` to customize your carousel [#25](https://github.com/sheikalthaf/ngx-carousel/issues/25)

* Fixed the back button issue when remaining slide is one [#28](https://github.com/sheikalthaf/ngx-carousel/issues/28)

* Point is no show the correct numbers when resize [#29](https://github.com/sheikalthaf/ngx-carousel/issues/29)

* button will gets hidden when no more slides [#33](https://github.com/sheikalthaf/ngx-carousel/issues/33)

* This feature is available through `onMove` output. check `visibleItems` [#34](https://github.com/sheikalthaf/ngx-carousel/issues/33)

* point click event is now added. if you click any point in move to curresponding slide[#17](https://github.com/sheikalthaf/ngx-carousel/issues/17)

## 1.2.5

### Features

* In inputs `point` is now changed with more options (`point: {visible: boolean, pointStyles: string}`).

* In point `visible` is used to visible point in the carousel.

* In point `pointStyles` is used to customize the point with own css. Check readme file for more details.

* Point click event is now added. If you click any point in move to curresponding slide

### Issues fixed

* custom css option is added in `pointStyles`[#16](https://github.com/sheikalthaf/ngx-carousel/issues/16)

* point click event is now added. if you click any point in move to curresponding slide[#17](https://github.com/sheikalthaf/ngx-carousel/issues/17)

## 1.2.0

### Changes

* From this build `inputsLength` Input and `dynamicLength` is removed.

* Added a new directive for the items `NgxCarouselItem` and it is mandatory. Check getstarted in readme.

* Added a new directive for buttons `NgxCarouselNext` and `NgxCarouselPrev`. it is mandatory. Check getstarted in readme.

* Removed buttons css for user customization. From this build `NgxCarouselNext` act as a rightRs and `NgxCarouselPrev` act as a leftRs. go and apply your own button with these directives. Check getstarted in readme.

### Issues fixed

* Example is Updated in readme[#13](https://github.com/sheikalthaf/ngx-carousel/issues/13)

* buttons issue is fixed.[#12](https://github.com/sheikalthaf/ngx-carousel/issues/12)

* `ExpressionChangedAfterItHasBeenCheckedError` issue is fixed.[#11](https://github.com/sheikalthaf/ngx-carousel/issues/11)

* classList of undefined is fixed[#10](https://github.com/sheikalthaf/ngx-carousel/issues/10)

* touch slideing issue is fixed[#4](https://github.com/sheikalthaf/ngx-carousel/issues/4)

## 1.1.1

### Changes

* `CarouselModule` is renamed to `NgxCarouselModule` to match the name of the caroussel. Please change the name.
* `Carousel` interface is renamed to `NgxCarousel` to match the name of the caroussel. Please change the name.

## 1.1.0

### Features Added

* `loop: boolean` is added. By default it is false. It is used to control the looping of an item. If you need `interval: boolean` then make sure to to add `loop: true`. [#5](https://github.com/sheikalthaf/ngx-carousel/issues/5)

* `ngx-item` component is added for the replacement of the `.item` class. check readme.

* `ngx-tile` component is added for the replacement of the `.item` and `.tile` class. check readme.

* `easing: string` is added for custom easing style.

* Performance Improvements.

### RoadMap

* support for Angular Universal
* Custom Animation support
* Unsubcribe to carouselLoad Eventemitter

## 1.0.0

### Issues fixed

* Buttton error fixed based on the issue[#2](https://github.com/sheikalthaf/ngx-carousel/issues/2)

### Features Added

* This release now `all` feature is enabled. It will appear in `grid`. if you need to make your carousel items to be `fixed in width` use this option and makke other option to be `0`. eg `grid: {xs: 0, sm: 0, md: 0, lg: 0, all: 250}`

* touch control option included in this version, By default touch is false.
* Performance Improvements.

## 0.2.2

### This is a Major release with breaking Changes. Please read the [Readme.md](https://github.com/sheikalthaf/ngx-carousel/blob/master/README.md) before upgrading


### Features Added

* This release now supports angular 2 projects (previously it requires angular 4).
* Performance Improvements



## 0.2.0

### This is a Major release with breaking Changes. Please read the [Readme.md](https://github.com/sheikalthaf/ngx-carousel/blob/master/README.md) before upgrading

### This is a major release with api changes


src/carousel.directive.ts
import { Directive, ElementRef, Renderer, Input, Output, OnInit, HostListener, EventEmitter } from '@angular/core';

@Directive({
  // tslint:disable-next-line:directive-selector
  selector: '[carousel]'
})
export class CarouselDirective {

    @Input() inputs: any;
    @Output() carouselLoad: EventEmitter<any> = new EventEmitter();

    private carousel: any;
    private carouselInner: any;
    private carouselItems: any;
    private onResize: any;
    private ItemsDetector: any;
    private onScrolling: any;
    private itemPosition: any;
    private carouselData: any = {
      'width': 0,
      'visibleItems': 0,
      'slideItems': 0
    }

    // private itemWidth: number;
    private currentSlide: number;
    // private visibleItems: number;
    // private slideItems: number;

    private styleManager: any = {
      'style': ' .ngxcarousel {width: 100%;position: relative;} .ngxcarousel-inner {overflow-x: hidden;white-space: nowrap;font-size: 0;vertical-align: top;} .ngxcarousel-inner .item {display: inline-block;font-size: 14px;white-space: initial;}.leftRs {position: absolute;margin: auto;top: 0;bottom: 0;z-index: 100;left: 0;width: 50px;height: 50px;box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);border-radius: 999px;}.rightRs {position: absolute; margin: auto; top: 0; right: 0; bottom: 0;z-index: 100; width: 50px;height: 50px; box-shadow: 1px 2px 10px -1px rgba(0, 0, 0, .3);border-radius: 999px;}',
      'point': '.ngxcarouselPoint ul {list-style-type: none;text-align: center;padding: 12px;margin: 0;white-space: nowrap;overflow: auto;} .ngxcarouselPoint li {display: inline-block;border-radius: 50%;background: rgba(0, 0, 0, 0.55);padding: 4px;margin: 0 4px;transition-timing-function: cubic-bezier(.17, .67, .83, .67);transition: .4s;} .ngxcarouselPoint li.active {background: #6b6b6b;transform: scale(1.8);}',
      'animation': {'lazy': '.ngxcarouselLazy .item {transition: .6s ease all;}'},
      'customCss': {'tile':'.ngxcarousel-inner .item .tile {background: white;box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);margin: 5px;}',
                    'banner': {'point': '.banner .ngxcarouselPoint li {background: rgba(255, 255, 255, 0.55);}.banner .ngxcarouselPoint li.active {  background: white;}.banner .ngxcarouselPoint {    position: absolute;    width: 100%;    bottom: 20px; }'
                  }
      }
    };



    constructor(
      private el: ElementRef,
      private renderer: Renderer) {
      // renderer.setElementStyle(el.nativeElement, 'background', 'grey');
    }

    // tslint:disable-next-line:use-life-cycle-interface
    ngOnInit() {
      // let t0 = performance.now();
      this.carousel = this.el.nativeElement;
      this.carouselInner = this.carousel.getElementsByClassName('ngxcarousel-inner')[0];
      this.carouselItems = this.carouselInner.getElementsByClassName('item');
      this.carousel.querySelector('.rightRs').addEventListener('click', () => { this.carouselScroll(1); });
      this.carousel.querySelector('.leftRs').addEventListener('click', () => { this.carouselScroll(0); });
      const itemsSplit = this.inputs.items.split('-');
        this.inputs.items = {
          'xs': itemsSplit[0],
          'sm': itemsSplit[1],
          'md': itemsSplit[2],
          'lg': itemsSplit[3]
        }
      this.carouselSize()
      this.storeResData();
      if (this.inputs.point === true) {
          this.carouselPoint();
          this.carouselInner.onscroll = () => {
            this.scrollStart();
          }
          this.carouselInner.addEventListener('DOMNodeInserted', () => {
          clearTimeout(this.ItemsDetector);
          this.ItemsDetector = setTimeout(() => { this.carouselPoint() }, 500);
        });
      }
      // let t1 = performance.now();
      // console.log('Took', (t1 - t0).toFixed(4), 'milliseconds to generate');
    }

    @HostListener('mouseenter')
    onmouseenter() {
       this.renderer.setElementClass(this.carousel, 'ngxcarouselHovered', true);
    }

    @HostListener('mouseleave')
    onmouseleave() {
       this.renderer.setElementClass(this.carousel, 'ngxcarouselHovered', false);
    }

    @HostListener('window:resize')
    onResizing() {
      clearTimeout(this.onResize);
      this.onResize = setTimeout(() => { this.storeResData(); }, 500);
    }

    @HostListener('window:scroll')
    onWindowScrolling() {
      clearTimeout(this.onScrolling);
      this.onScrolling = setTimeout(() => {
        const top = this.carousel.offsetTop;
        const scrollY = window.scrollY;
        const heightt = window.innerHeight;
        const carouselHeight = this.carousel.offsetHeight;
        if ((top <= scrollY + heightt - carouselHeight / 4) && (top + carouselHeight / 2 >= scrollY)) {
          this.renderer.setElementClass(this.el.nativeElement, 'ngxcarouselScrolled', false);
        } else {
          this.renderer.setElementClass(this.el.nativeElement, 'ngxcarouselScrolled', true);
        }
      }, 800);

    }

    private scrollStart() {
      clearTimeout(this.itemPosition);
      this.itemPosition = setTimeout(() => {
        const scroll = this.carouselInner.scrollLeft;
        const itm = Math.round(scroll / this.carouselData.width);
        const left = itm * this.carouselData.width;
        // tslint:disable-next-line:no-unused-expression
        scroll !== left && this.smoothScoll(left, 200);
        this.carouselPointActive();
      }, 400);
    }

    private storeResData() {
        const w = window.innerWidth;
        // tslint:disable-next-line:max-line-length
        this.carouselData.items = +(w >= 1200 ? this.inputs.items.lg : w >= 992 ? this.inputs.items.md : w >= 768 ? this.inputs.items.sm : this.inputs.items.xs);
        this.carouselData.width = this.carouselInner.offsetWidth / this.carouselData.items;
        this.carouselInner.scrollLeft = this.currentSlide * this.carouselData.width;
        this.carouselData.slideItems = +(this.inputs.slide < this.carouselData.items ? this.inputs.slide : this.carouselData.items);
        this.inputs.speed = +(this.inputs.speed ? this.inputs.speed : 400);
    }

    private carouselPoint() {
        const Nos = this.carouselItems.length - (this.carouselData.items - this.carouselData.slideItems);
        const p = Math.ceil(Nos / this.carouselData.slideItems);
        console.log(this.carouselData.items +","+ this.carouselData.slideItems);
        // tslint:disable-next-line:max-line-length
        // tslint:disable-next-line:no-unused-expression
        this.carousel.querySelectorAll('.ngxcarouselPoint').length === 0 && this.carousel.insertAdjacentHTML('beforeend', '<div class="ngxcarouselPoint"></div>')
        let point = '<ul>'
        for (let i = 0; i < p; i++) {
                point += '<li></li>';
        }
        point += '</ul>';
        const respo = this.carousel.querySelectorAll('.ngxcarouselPoint')[0];
        respo.innerHTML = '';
        respo.insertAdjacentHTML('beforeend', point);
        this.carouselPointActive();
    }

    private carouselPointActive() {
        const parent = this.carousel.querySelectorAll('.ngxcarouselPoint li');
        if ( parent.length !== 0) {
          const start = Math.round(this.carouselInner.scrollLeft / this.carouselData.width);
          const i = Math.ceil(start / this.carouselData.slideItems);
          for (let j = 0; j < parent.length; j++) {
              this.renderer.setElementClass(parent[j], 'active', false);
          }
          this.renderer.setElementClass(parent[i], 'active', true);
        }
    }

    private carouselSize() {
        const t0 = performance.now();
        // const index = + Math.round( Math.random() * (1000 - 10) + 10);
        const id = 'ResSlid' + this.makeid();

        const styleCollector0: string = '.' + id + ' .item {width: ' + 100 / this.inputs.items.xs + '%}';
        const styleCollector1: string = '.' + id + ' .item {width: ' + 100 / this.inputs.items.sm + '%}';
        const styleCollector2: string = '.' + id + ' .item {width: ' + 100 / this.inputs.items.md + '%}';
        const styleCollector3: string = '.' + id + ' .item {width: ' + 100 / this.inputs.items.lg + '%}';

        let dism = '';
        if (typeof this.inputs.custom !== 'undefined' && this.inputs.custom === 'banner') {
          dism += this.styleManager.customCss.banner.point
        }
        if (typeof this.inputs.custom !== 'undefined' && this.inputs.custom === 'tile') {
          dism += this.styleManager.customCss.tile
        }
        if (typeof this.inputs.animation !== 'undefined' && this.inputs.animation === 'lazy') {
          dism += this.styleManager.animation.lazy
          this.renderer.setElementClass(this.carousel, 'ngxcarouselLazy', true);
        }

        const styleCollector = '<div><style>' +
        this.styleManager.style +
        this.styleManager.point +
        dism +
                      '@media (max-width:767px){.ngxcarousel-inner {overflow-x: auto; }' + styleCollector0 + '}' +
                      '@media (min-width:768px){' + styleCollector1 + '}' +
                      '@media (min-width:992px){' + styleCollector2 + '}' +
                      '@media (min-width:1200px){' + styleCollector3 + '}</style></div>';

        this.renderer.setElementClass(this.carousel, id, true);
        this.carousel.insertAdjacentHTML('beforeend', styleCollector);
        this.carouselSlide();
        setTimeout(() => {
          this.storeResData();
        }, 20);
        const t1 = performance.now();
        console.log('Took', (t1 - t0).toFixed(4), 'milliseconds to generate');
    }

    private makeid() {
        let text = '';
        const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

        for (let i = 0; i < 5; i++ ) {
            text += possible.charAt(Math.floor(Math.random() * possible.length));
        }
        return text;
    }

    private carouselSlide() {
      if (this.inputs.interval !== '') {
        setInterval(() => {
    		  // tslint:disable-next-line:max-line-length
    		  // tslint:disable-next-line:no-unused-expression
           !(this.carousel.classList.contains('ngxcarouselHovered')) && !(this.carousel.classList.contains('ngxcarouselScrolled')) && this.carouselScroll(1);
        }, +(this.inputs.interval));
      }
    }

    private carouselScroll(Btn: number) {
        const t0 = performance.now();
        let itemSpeed = +this.inputs.speed;
        let translateXval, currentSlide = 0;
        const itemLenght = this.carouselItems.length;
        const divValue = Math.round(this.carouselInner.scrollLeft / this.carouselData.width);

        if (Btn === 0) {
            currentSlide = divValue - this.carouselData.slideItems;
            translateXval = currentSlide * this.carouselData.width;
            const MoveSlide = currentSlide + this.carouselData.slideItems;

            if (divValue === 0) {
                currentSlide = itemLenght - this.carouselData.slideItems;
                translateXval = currentSlide * this.carouselData.width;
                currentSlide = itemLenght - this.carouselData.items;
                itemSpeed = 400;
                Btn = 1;
            } else if (this.carouselData.slideItems >= MoveSlide) {
                currentSlide = translateXval = 0;
            }
        } else {
            currentSlide = divValue + this.carouselData.slideItems;
            translateXval = currentSlide * this.carouselData.width;
            const MoveSlide = currentSlide + this.carouselData.slideItems;

            if (divValue + this.carouselData.items === itemLenght) {
                currentSlide = translateXval = 0;
                itemSpeed = 400;
                Btn = 0;
            } else if (itemLenght <= (MoveSlide - this.carouselData.slideItems + this.carouselData.items)) {
                currentSlide = itemLenght - this.carouselData.slideItems;
                translateXval = currentSlide * this.carouselData.width;
                currentSlide = itemLenght - this.carouselData.items;
            }
        }
        // console.log(this.carouselInner.offsetWidth +","+ this.carouselData.items +","+ translateXval +","+ this.carouselData.width);
        // tslint:disable-next-line:max-line-length
        // tslint:disable-next-line:no-unused-expression
        this.inputs.animation === 'lazy' && this.carouselAnimator( Btn, currentSlide + 1, currentSlide + this.carouselData.items, itemSpeed, Math.abs(divValue - currentSlide));
        this.currentSlide = currentSlide;
        this.carouselLoad1();
        this.smoothScoll(translateXval, itemSpeed);

        const t1 = performance.now();
        // console.log('Took', (t1 - t0).toFixed(4), 'milliseconds to generate');
    }

    private carouselLoad1() {
      if (this.inputs.load !== '') {
        this.inputs.load = this.inputs.load >= this.carouselData.slideItems ? this.inputs.load : this.carouselData.slideItems;
        // tslint:disable-next-line:max-line-length
        // tslint:disable-next-line:no-unused-expression
        (this.carouselItems.length - this.inputs.load) <= (this.currentSlide + this.carouselData.items) && this.carouselLoad.emit(this.currentSlide);
      }
    }

    private easeinq(t: number) {
      // return t<.5 ? 2*t*t : -1+(4-2*t)*t;
    }

    private smoothScoll(travel: number, speed: number) {
        // console.log(travel);
        const startX = this.carouselInner.scrollLeft;
        const stopX = travel;
        const distance = stopX > startX ? stopX - startX : startX - stopX;
        // let speed = Math.round(distance / 100);
        //    if (speed >= 20) speed = 20;
        /// const rol = Math.round(distance/speed)*10+10;
        const rol = 13;
        // console.log(rol);
        const step = Math.round(speed / rol);
        const addr = distance / step;
        // let leapX = stopX > startX ? startX + addr : startX - addr;
        let leapX = startX;
          if (stopX > startX) {
              for (let i = 0; i <= speed; i += rol) {
                setTimeout(() => {
                    // console.log(this.easeinq(leapX)+","+leapX);
                    leapX += addr
                    if (Math.round(leapX) + 5 >= stopX) { leapX = stopX;   }
                    this.carouselInner.scrollLeft = Math.round(leapX);
                    if (leapX === stopX) {return}
                  }, i);
              }
          } else {
              for (let i = 0; i <= speed; i += rol) {
                setTimeout(() => {
                    leapX -= addr
                    if (Math.round(leapX) - 5 <= stopX) { leapX = stopX;  }
                    this.carouselInner.scrollLeft = Math.round(leapX);
                    if (leapX === stopX) {return}
                  }, i);
              }
          }
    }

    private carouselAnimator( direction: number, start: number, end: number, speed: number, length: number) {
        let val = length < 5 ? length : 5;
            val = val === 1 ? 3 : val;

        if (direction === 1) {
          for (let i = start - 1; i < end; i++) {
              val = val * 2;
              this.carouselItems[i].style.transform = 'translateX(' + val + 'px)';
          }
        } else {
          for (let i = end - 1; i >= start - 1; i--) {
              val = val * 2;
               this.carouselItems[i].style.transform = 'translateX(-' + val + 'px)';
          }
        }
        setTimeout(() => {
          for (let i = start - 1; i < end; i++) {
            this.carouselItems[i].style.transform = 'translateX(0px)';
          }
        }, speed - 70);
    }

}
